src/service/video-manager-impl.js

```diff
@@ -959,6 +932,369 @@ class VideoEntry {
+
+
+/** Manages rotate-to-fullscreen video. */
+class R82FullscreenManager {
+
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+    const {win} = ampdoc;
+
+    /** @private @const {!./ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @private {number} */
+    this.nextId_ = 0;
+
+    /** @private @const {!Window} */
+    this.win_ = win;
+
+    /** @private @const {!./timer-impl.Timer} */
+    this.timer_ = Services.timerFor(win);
+
+    /** @private {?R82FullscreenEntryDef} */
+    this.currently_ = null;
+
+    /**
+     * Maps rotate-to-fullscreen entry id to entry.
+     * @private @const {!Object<string, !R82FullscreenEntryDef>}
+     */
+    this.entries_ = {};
+
+    /** @private @const {function()} */
+    const installOrientationObserver =
+        this.installOrientationObserver_.bind(this);
+    this.installOrientationObserver_ = once(installOrientationObserver);
+
+    /** @private @const {function():!IntersectionObserver} */
+    const getIntersectionObserver = this.getIntersectionObserver_.bind(this);
+    this.getIntersectionObserver_ =
+      /** @type {function():!IntersectionObserver} */ (
+        once(getIntersectionObserver));
+  }
+
+  /**
+   * @param {!AmpElement} element
+   * @param {!VideoEntry} videoEntry
+   */
+  register(element, videoEntry) {
+    const id = this.nextId_++;
+
+    element[R82_FULLSCREEN_ID_PROP] = id;
+
+    element.getImpl().then(video => {
+      this.entries_[id.toString()] = {videoEntry, video};
+    });
+
+    this.installOrientationObserver_();
+
+    // Wait until the video is "blessed" in order to have fullscreen access.
+    element.signals().whenSignal(Signals.BLESSED).then(() => {
+      this.getIntersectionObserver_().observe(element);
+      if (!this.isLandscape_()) {
+        return;
+      }
+      const {video} = videoEntry;
+      const videoForVsync = /** @type {!../base-element.BaseElement} */ (video);
+      this.updateVisibilityOnBless_(videoForVsync);
+    });
+  }
+
+  /** @private */
+  // Overridden in constructor
+  installOrientationObserver_() {
+    // TODO(alanorozco) Update based on support
+    const win = this.win_;
+    const screen = win.screen;
+    // Chrome considers 'orientationchange' to be an untrusted event, but
+    // 'change' on screen.orientation is considered a user interaction.
+    // We still need to listen to 'orientationchange' on Chrome in order to
+    // exit fullscreen since 'change' does not fire in this case.
+    if (screen && 'orientation' in screen) {
+      const orient = /** @type {!ScreenOrientation} */ (screen.orientation);
+      listen(orient, 'change', () => this.onRotation_());
+    }
+    // iOS Safari does not have screen.orientation but classifies
+    // 'orientationchange' as a user interaction.
+    listen(win, 'orientationchange', () => this.onRotation_());
+  }
+
+  /** @private */
+  onRotation_() {
+    const root = this.ampdoc_.getRootNode();
```

Comment:
do, instead of using intersectionObserver, can we just call `getIntersectionChangeEntry` on video elements and calculate based on that?

If `orientation` event happens too late ( e.g. after `getIntersectionChangeEntry` values have been updated based on landscape, then you can do the calculations in a `viewer.onScroll` event. All `getIntersectionChangeEntry` values are cached by runtime so there is no perf hit. `autoplay` does it this way)